File: extract_data.py
#!/usr/bin/env python3
"""
Extract all social media data from the HTML dashboard file
"""

import re
import json
from html import unescape
from typing import Dict, List, Any

def extract_all_data():
    """Extract all card data from the HTML file"""
    
    # Read the entire file
    with open('/Users/boshaffer/Downloads/Campaign 2/intel-dashboard-vercel/public/dashboard.html', 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Find all article elements with their complete content
    article_pattern = r'<article class="card"[^>]*>(.*?)</article>'
    articles = re.findall(article_pattern, content, re.DOTALL)
    
    extracted_data = []
    
    for article_html in articles:
        # Extract the opening tag to get data attributes
        full_article = f'<article class="card"{article_html.split(">")[0]}>'
        
        # Extract data attributes
        data_kind = re.search(r'data-kind="([^"]*)"', full_article)
        data_lean = re.search(r'data-lean="([^"]*)"', full_article)
        data_cands = re.search(r'data-cands="([^"]*)"', full_article)
        data_issues = re.search(r'data-issues="([^"]*)"', full_article)
        data_engagement = re.search(r'data-engagement="([^"]*)"', full_article)
        data_text = re.search(r'data-text="([^"]*)"', full_article)
        
        # Extract time/author from card-head
        time_match = re.search(r'<span class="time">([^<]*)</span>', article_html)
        author_match = re.search(r'<span class="time">@([^<]*)</span>', article_html)
        
        # Extract URL from external link
        url_match = re.search(r'<a class="ext" href="([^"]*)"', article_html)
        
        # Extract text content
        text_section = re.search(r'<div class="text">(.*?)</div>', article_html, re.DOTALL)
        text_content = ""
        if text_section:
            text_raw = text_section.group(1)
            # Remove HTML tags and decode HTML entities
            text_content = re.sub(r'<[^>]+>', '', text_raw)
            text_content = unescape(text_content).strip()
        
        # Extract engagement metrics
        meta_section = re.search(r'<div class="meta">(.*?)</div>', article_html, re.DOTALL)
        engagement_metrics = {}
        if meta_section:
            likes = re.search(r'❤ <b>([^<]*)</b>', meta_section.group(1))
            retweets = re.search(r'↻ <b>([^<]*)</b>', meta_section.group(1))
            comments = re.search(r'💬 <b>([^<]*)</b>', meta_section.group(1))
            views = re.search(r'👁 <b>([^<]*)</b>', meta_section.group(1))
            
            if likes:
                engagement_metrics['likes'] = likes.group(1)
            if retweets:
                engagement_metrics['retweets'] = retweets.group(1)
            if comments:
                engagement_metrics['comments'] = comments.group(1)
            if views:
                engagement_metrics['views'] = views.group(1)
        
        # Extract opposition research notes if present
        oppo_note = ""
        oppo_match = re.search(r'<div class="oppo-note">(.*?)</div>', article_html, re.DOTALL)
        if oppo_match:
            oppo_note = re.sub(r'<[^>]+>', '', oppo_match.group(1))
            oppo_note = unescape(oppo_note).strip()
        
        # Build the data object
        item = {
            'id': len(extracted_data) + 1,
            'kind': data_kind.group(1) if data_kind else '',
            'lean': data_lean.group(1) if data_lean else '',
            'candidates': data_cands.group(1).split(',') if data_cands and data_cands.group(1) else [],
            'issues': data_issues.group(1).split(',') if data_issues and data_issues.group(1) else [],
            'engagement_score': int(data_engagement.group(1)) if data_engagement and data_engagement.group(1).isdigit() else 0,
            'data_text': unescape(data_text.group(1)) if data_text else '',
            'text_content': text_content,
            'timestamp': time_match.group(1) if time_match else '',
            'author': author_match.group(1) if author_match else '',
            'url': url_match.group(1) if url_match else '',
            'engagement_metrics': engagement_metrics,
            'opposition_note': oppo_note
        }
        
        extracted_data.append(item)
    
    return extracted_data

def categorize_data(data):
    """Categorize data by type and source"""
    categories = {
        'twitter_mentions': [],
        'twitter_posts': [],
        'reddit_posts': [],
        'reddit_comments': [],
        'tiktok_videos': [],
        'instagram_posts': [],
        'instagram_videos': [],
        'instagram_carousels': [],
        'opposition_research': [],
        'poll_data': []
    }
    
    for item in data:
        kind = item['kind']
        
        if 'oppo' in item.get('url', '') or item['opposition_note']:
            categories['opposition_research'].append(item)
        elif 'poll' in item['text_content'].lower() or 'favorability' in item['text_content'].lower():
            categories['poll_data'].append(item)
        elif kind == 'post' and 'x.com' in item.get('url', ''):
            if any(cand in item['text_content'].lower() for cand in ['johnson', 'brewer', 'giannoulias']):
                categories['twitter_mentions'].append(item)
            else:
                categories['twitter_posts'].append(item)
        elif kind == 'reply' and 'x.com' in item.get('url', ''):
            categories['twitter_mentions'].append(item)
        elif kind == 'post' and 'reddit' in item.get('url', ''):
            categories['reddit_posts'].append(item)
        elif kind == 'comment':
            categories['reddit_comments'].append(item)
        elif kind == 'video' and 'tiktok' in item.get('url', ''):
            categories['tiktok_videos'].append(item)
        elif kind == 'photo':
            categories['instagram_posts'].append(item)
        elif kind == 'video' and 'instagram' in item.get('url', ''):
            categories['instagram_videos'].append(item)
        elif kind == 'carousel':
            categories['instagram_carousels'].append(item)
    
    return categories

if __name__ == "__main__":
    print("Extracting all data from dashboard.html...")
    
    # Extract all data
    all_data = extract_all_data()
    print(f"Extracted {len(all_data)} total items")
    
    # Categorize the data
    categorized = categorize_data(all_data)
    
    # Print summary
    print("\nData summary by category:")
    for category, items in categorized.items():
        print(f"{category}: {len(items)} items")
    
    # Save to JSON for verification
    with open('/Users/boshaffer/Downloads/Campaign 2/intel-dashboard-vercel/extracted_data.json', 'w', encoding='utf-8') as f:
        json.dump({
            'all_data': all_data,
            'categorized': categorized,
            'total_count': len(all_data)
        }, f, indent=2, ensure_ascii=False)
    
    print(f"\nData exported to extracted_data.json")
    print("Ready to generate TypeScript file...")

File: scripts/extractData.js
const fs = require('fs');
const path = require('path');

// Read the HTML file
const htmlPath = path.join(__dirname, '../public/dashboard.html');
const html = fs.readFileSync(htmlPath, 'utf8');

// Extract all card data using regex
const cardRegex = /<article class="card"[^>]*>([\s\S]*?)<\/article>/g;
const cards = [];
let match;

while ((match = cardRegex.exec(html)) !== null) {
  const cardHtml = match[0];
  
  // Extract data attributes
  const dataKind = (cardHtml.match(/data-kind="([^"]*)"/) || [])[1];
  const dataLean = (cardHtml.match(/data-lean="([^"]*)"/) || [])[1];
  const dataCands = (cardHtml.match(/data-cands="([^"]*)"/) || [])[1];
  const dataIssues = (cardHtml.match(/data-issues="([^"]*)"/) || [])[1];
  const dataEngagement = (cardHtml.match(/data-engagement="([^"]*)"/) || [])[1];
  const dataText = (cardHtml.match(/data-text="([^"]*)"/) || [])[1];
  
  // Extract badges
  const badges = [];
  const badgeMatches = cardHtml.matchAll(/<span class="badge ([^"]*)"[^>]*>([^<]*)<\/span>/g);
  for (const badgeMatch of badgeMatches) {
    if (!badgeMatch[1].includes('cand') && !badgeMatch[1].includes('issue')) {
      badges.push({
        type: badgeMatch[1],
        text: badgeMatch[2]
      });
    }
  }
  
  // Extract time and author
  const timeMatch = cardHtml.match(/<span class="time">([^<]*)<\/span>/);
  const date = timeMatch ? timeMatch[1] : '';
  const authorMatch = cardHtml.match(/<span class="time">@([^<]*)<\/span>/);
  const author = authorMatch ? authorMatch[1] : '';
  
  // Extract URL
  const urlMatch = cardHtml.match(/href="(https:\/\/[^"]*)"[^>]*rel="noopener"/);
  const url = urlMatch ? urlMatch[1] : '';
  
  // Extract text content
  const textMatch = cardHtml.match(/<div class="text">([^<]*(?:<a[^>]*>[^<]*<\/a>[^<]*)*)<\/div>/);
  const text = textMatch ? textMatch[1].replace(/<[^>]*>/g, '').replace(/&amp;/g, '&').replace(/&#x27;/g, "'") : '';
  
  // Extract engagement metrics
  const engagement = {};
  if (cardHtml.includes('❤')) {
    const likesMatch = cardHtml.match(/❤\s*<b>(\d+)<\/b>/);
    engagement.likes = likesMatch ? parseInt(likesMatch[1]) : 0;
  }
  if (cardHtml.includes('↻')) {
    const retweetsMatch = cardHtml.match(/↻\s*<b>(\d+)<\/b>/);
    engagement.retweets = retweetsMatch ? parseInt(retweetsMatch[1]) : 0;
  }
  if (cardHtml.includes('💬')) {
    const commentsMatch = cardHtml.match(/💬\s*<b>(\d+)<\/b>/);
    engagement.comments = commentsMatch ? parseInt(commentsMatch[1]) : 0;
  }
  if (cardHtml.includes('⬆')) {
    const upvotesMatch = cardHtml.match(/⬆\s*<b>(\d+)<\/b>/);
    engagement.upvotes = upvotesMatch ? parseInt(upvotesMatch[1]) : 0;
  }
  
  // Determine platform
  let platform = 'unknown';
  if (url.includes('x.com') || url.includes('twitter.com')) platform = 'twitter';
  else if (url.includes('reddit.com')) platform = 'reddit';
  else if (url.includes('tiktok.com')) platform = 'tiktok';
  else if (url.includes('instagram.com')) platform = 'instagram';
  else if (badges.some(b => b.text.includes('Reddit'))) platform = 'reddit';
  else if (badges.some(b => b.text.includes('X '))) platform = 'twitter';
  else if (badges.some(b => b.text.includes('TikTok'))) platform = 'tiktok';
  else if (badges.some(b => b.text.includes('IG'))) platform = 'instagram';
  
  cards.push({
    id: `card-${cards.length + 1}`,
    platform,
    type: dataKind || 'post',
    sentiment: dataLean || 'neutral',
    candidates: dataCands ? dataCands.split(',') : [],
    issues: dataIssues ? dataIssues.split(',') : [],
    engagementScore: parseInt(dataEngagement) || 0,
    date,
    author,
    url,
    text,
    engagement,
    badges
  });
}

// Extract opposition research sections
const oppoRegex = /<div class="intel-section[^"]*"[^>]*>([\s\S]*?)<\/div>\s*<\/div>/g;
const oppoSections = [];

while ((match = oppoRegex.exec(html)) !== null) {
  const sectionHtml = match[0];
  const titleMatch = sectionHtml.match(/<h2[^>]*>([^<]*)<\/h2>/);
  const contentMatch = sectionHtml.match(/<div class="intel-body">([^]*)<\/div>/);
  
  if (titleMatch && contentMatch) {
    oppoSections.push({
      id: `oppo-${oppoSections.length + 1}`,
      title: titleMatch[1],
      content: contentMatch[1].replace(/<[^>]*>/g, '').replace(/&amp;/g, '&')
    });
  }
}

// Group by platform
const byPlatform = {
  twitter: cards.filter(c => c.platform === 'twitter'),
  reddit: cards.filter(c => c.platform === 'reddit'),
  tiktok: cards.filter(c => c.platform === 'tiktok'),
  instagram: cards.filter(c => c.platform === 'instagram'),
  unknown: cards.filter(c => c.platform === 'unknown')
};

// Group by candidate
const byCandidates = {};
['brewer', 'johnson', 'quigley', 'pappas', 'mendoza', 'giannoulias'].forEach(cand => {
  byCandidates[cand] = cards.filter(c => 
    c.candidates.some(candidate => candidate.toLowerCase().includes(cand))
  );
});

// Calculate statistics
const stats = {
  total: cards.length,
  byPlatform: Object.entries(byPlatform).map(([platform, items]) => ({
    platform,
    count: items.length
  })),
  bySentiment: {
    positive: cards.filter(c => c.sentiment === 'pos').length,
    negative: cards.filter(c => c.sentiment === 'neg').length,
    mixed: cards.filter(c => c.sentiment === 'mixed').length,
    neutral: cards.filter(c => !c.sentiment || c.sentiment === 'neutral').length
  },
  byCandidate: Object.entries(byCandidates).map(([candidate, items]) => ({
    candidate,
    count: items.length
  })),
  topEngagement: cards.sort((a, b) => b.engagementScore - a.engagementScore).slice(0, 10)
};

// Save to JSON file
const output = {
  metadata: {
    extractedAt: new Date().toISOString(),
    totalCards: cards.length,
    oppositionSections: oppoSections.length
  },
  stats,
  cards,
  oppositionResearch: oppoSections,
  byPlatform,
  byCandidates
};

fs.writeFileSync(
  path.join(__dirname, '../lib/scrapedData.json'),
  JSON.stringify(output, null, 2)
);

console.log(`✅ Extracted ${cards.length} cards from dashboard`);
console.log(`📊 Platform breakdown:`, stats.byPlatform);
console.log(`📈 Sentiment:`, stats.bySentiment);
console.log(`🎯 By candidate:`, stats.byCandidate);
console.log(`📁 Data saved to lib/scrapedData.json`);